src/borg/archiver.py

```diff
@@ -687,91 +747,41 @@ def _process(self, *, path, parent_fd=None, name=None,
                             # if we are already recursing in an excluded dir, we do not need to do anything else than
                             # returning (we do not need to archive or recurse into tagged directories), see #3991:
                             if not recurse_excluded_dir:
-                                if keep_exclude_tags and not dry_run:
-                                    fso.process_dir(path=path, fd=child_fd, st=st)
+                                if keep_exclude_tags:
+                                    status = self._process_any(path=path, parent_fd=parent_fd, name=name, st=st, fso=fso,
+                                                          cache=cache, read_special=read_special, dry_run=dry_run)
```

Comment:
notable change:

we have locally opened (see `OsOpen` call here) the directory and have `child_fd` for it. the old code used that `child_fd` to process the directory, while the new code re-opens the directory again (see `OsOpen` call in `process_dir`).

guess we could also pass `fd=child_fd` to `process_any` and finally to `process_dir` to
- avoid a race condition here
- maybe also being useful for processing of stdin?

Reply:
Actually there's no real reason for this, i should revert this snippet (adding a `fso.process_dir_with_fd` function that is the same as the old `fso.process_dir`). It would've only been worth if there wasn't the tag-file exclusion thing (then `self._process_any` could've been called only once without special-casing directories).

Reply:
I pushed the revert without waiting for answer, will undo if you think it was better before. It does duplicate process_dir, but having it that way is more orthogonal and will be used for my "read-paths-from-stdin".

Reply:
looks good.

the reading from stdin needs some special casing anyway because it is not a fs file, so we can't read attrs / xattrs / etc. from the filesystem